### PR TITLE
Add `has_more_values` flag to fieldvalues

### PR DESCRIPTION
**Pull request by qnkhuat**
**Background**
When we analyze db, we classify which fields we could store a list of values in `metabase_fieldvalues` and we have 2 criteria for that:
1. the number of distinct values must be less than 100
2. the total length of all distinct values must be less than a 10000

if a field satisfied both conditions we call it an `auto-list` field and automatically save a list of values into `metabase_fieldvalues`.

The benefit of having an `auto-list` field is that if we create a parameter on a dashboard and connect to that field, UI could show the choices as a drop-down and we also don't need to query to the DB.

But sometimes users just really want to use a dropdown widget for columns that are not satisfied with our conditions, and users could do that if they change the field setting in the Data model.

Once users set It and trigger a field scan, we will extend our condition to save to `metabase_fieldvalues` a little bit:
1. the number of distinct values must be less than 5000 ( this is the hard limit of what we query from db )
2. the total length of all distinct values must be less than 1000 ( the same as before )

The 1st condition raises 1 problem:  how does UI know if the values stored in` metabase_fieldvalues` are all distinct values of that field, or it's just a part of all possible values? 
Think of a field that has 10000 distinct values, but we only store 5000 of them because it's all we query.

So this PR adds a flag `has_more_values` to the fiedlvalues to solve that problem, If it's `true`, it means that the values we store in `metabase_fieldvalues` are just a subset of all possible values.

Epic #23162

### CODE REVIEW

File: src/metabase/models/field_values.clj [hunk 1]
```diff
@@ -150,20 +156,39 @@
...
+      ;; we need to unmark it as `auto-list`. Switch it to `has_field_values` = `nil` and delete the FieldValues;
+      ;; this will result in it getting a Search Widget in the UI when `has_field_values` is automatically inferred
+      ;; by the [[metabase.models.field/infer-has-field-values]] hydration function (see that namespace for more detailed
       ;; discussion)
       ;;
       ;; It would be nicer if we could do this in analysis where it gets marked `:auto-list` in the first place, but
       ;; Fingerprints don't get updated regularly enough that we could detect the sudden increase in cardinality in a
       ;; way that could make this work. Thus, we are stuck doing it here :(
-      (and (> (count values) auto-list-cardinality-threshold)
+      (and (or values-exceed-length?
```
Comment (qnkhuat) [inline src/metabase/models/field_values.clj:190]: this is actually a case we didn't catch before,
previously if an `auto-list` field suddenly has the total length of all distinct values more than 10000, we will skip this step.

the correct way to do it is we need to remove the field values and change `has_field_values` to `nil` so that UI will display a search widget for it.

-- I also added this case in the test

File: test/metabase/related_test.clj [hunk 2]
```diff
@@ -11,9 +11,9 @@
 (deftest collect-context-bearing-forms-test
   (is (= #{[:field 1 nil] [:metric 1] [:field 2 nil] [:segment 1]}
          (#'related/collect-context-bearing-forms [[:> [:field 1 nil] 3]
-                                             ["and" [:= ["FIELD-ID" 2] 2]
-                                              ["segment" 1]]
-                                             [:metric 1]]))))
+                                                   ["and" [:= ["FIELD-ID" 2] 2]
```
Comment (qnkhuat) [inline test/metabase/related_test.clj:14]: unrelated change

**Comment by dpsutton** [comment]
Did we consider changing the UI component? It seems heavy handed to require changes to the data_model screen just to be able to type in a choice. And this can easily come up in scenarios where there are only 3 choices. These are only found once a day (if field values sync is enabled). So new values can be missing from this list for up to 24 hours either way. 

There could be some merit to allowing typing in values regardless of how confident we are in enumerating the whole set of values (except for boolean i suppose).

File: resources/migrations/000_migrations.yaml [hunk 3]
```diff
@@ -11846,6 +11846,21 @@ databaseChangeLog:
...
+        - addColumn:
+            tableName: metabase_field
+            columns:
+              - column:
+                  name: exceeded_limit
+                  type: boolean
+                  remarks: Whether or not a field has values store in metabase_fieldvalues is exceeded the limits
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (camsaul) [inline resources/migrations/000_migrations.yaml:11863]: It seems weird to add a `exceeded_limit` column to Field to record whether or not its FieldValues have exceeded some limit... why not add this column to FieldValues?

Comment (camsaul) [reply resources/migrations/000_migrations.yaml:11863]: Failing that something like `field_values_exceeded_limit` would be a more appropriate name. If I see a Field with `{:exceeded_limit false}` I have no idea what that means unless I go read this migration here

**Comment by camsaul** [comment]
Dumb question but is this something we actually need to record or can we just determine whether the synced values were truncated by looking at whether the number of values = 5000 or the length of the values = 10000? We could calculate that as part of the Toucan `post-select` or something. It would work in every case except for when we have _exactly_ 5000 distinct values, but I think that's an acceptable trade off. Then we'd have one less thing to sync in the app DB. Less things synced in the app DB = less moving parts = less things that can break

**Comment by qnkhuat** [comment]
@dpsutton 
> It seems heavy handed to require changes to the data_model screen just to be able to type in a choice. And this can easily come up in scenarios where there are only 3 choices.

I'm not sure I fully understand your concern here, but for fields that meet our thresholds ( less than 100 choices, and total length < 10000), we automatically cache the list of values and display a dropdown on UI, users don't have to change anything in data_model. They only have to change if a field exceeds our limits.

>  So new values can be missing from this list for up to 24 hours either way.

Yes, this is a valid concern; but I think it's the trade-off we accept when we implement this 6? years ago.

FYI, we have a plan of reviewing fieldvalues related stuff in this [project](https://www.notion.so/metabase/Theme-field-values-fa6c3febab8b40cfbae99eab18474042), I'll make sure we take into account this problem.

**Comment by qnkhuat** [comment]
@camsaul 

I store this field on `Field` because for cases where the list of values exceeds the length limit, we don't store `FieldValues` at all ([ref](https://github.com/metabase/metabase/blob/d67d6f063e5169d9e78b3374f95bee8f429cb1fd/src/metabase/models/field_values.clj#L200)).

One possible solution is in case exceeding the length limit, we will only store up to `10000` characters, dropping everything else after that.
But we still need the `exceed_limit` flag in this case, because we can't tell from a list of values whether or not it's a full list or sub-list.

**Comment by qnkhuat** [comment]
Updates:
- I've moved the `exceeded_limit` flag on field to `has_more_values` on fieldvalues
- For field that has `has_field_values=:list` and the fetched values exceeded the 10k chars limit, we will store a subset of that list up to 10k chars in fieldvalues -- note that this is new behavior, previously we don't store at all.
  - this will also resolve #16102

**Comment by dpsutton** [comment]
This still doesn't feel super great to me. This notion of recognizing our stored values are just a subset is a good one, but I don't think we can recognize when it is the case and should be open to this being the case at all times.

The original statements are incorrect:

    1. the number of distinct values must be less than 100
    2. the total length of all distinct values must be less than a 10000


We mark things as auto-list and save them if there are fewer than 100 distinct values while fingerprinting. This has nothing to do with how many distinct values are in the table but how many are in some random 10,000 row sample. And I think this is what bothers me a bit about this change. We are building in an escape hatch so the app understands that what it considers the full set of values might not be correct. But this is the case almost all of the time: the fingerprint sample did not discover all values, new values have been added, etc. So why not open this up by default all the time and recognize that we have a handy list of values but we are never sure it is all of them and allow free input for those cases.

**Comment by qnkhuat** [comment]
> This has nothing to do with how many distinct values are in the table but how many are in some random 10,000 row sample. 

This is true, but the values of those auto-list fields will be later re-populated with the scanning process that runs daily.
And we use select distinct for that scanning process, so in case new values are added, the values in fieldvalues will be updated.

> So why not open this up by default all the time and recognize that we have a handy list of values but we are never sure it is all of them and allow free input for those cases.

Correction: for auto-list, the list of values should be correct as long as no values were being added in a 24 hours period.

For fields that are frequently updated and have new values added on daily basis, the list of values will become incorrect. But at the same time, I think those fields will grow to pass our limit and be removed from `auto-list`.

In that case, if users want to use the dropdown again and they can go to mark those fields as `list`, I think it's fine for us to cache a sublist of that field, and as user type, we make a search query to the database.

**Comment by luizarakaki** [comment]
The fingerprint process defines if a field should be scanned or not. But the scanning process does know with 100% confidence whether a field has more than 1000 distinct values or not.

We run `SELECT field GROUP BY field LIMIT 1000`. So we will only be wrong about it if the field gets more than 1k distinct values between two scanning processes.

As we run scans frequently, I believe it is a lesser problem (that we already have today) and we don't have many user complaints.

**Comment by dpsutton** [comment]
oh good point @luizarakaki i forgot there's a separate query and it is not just using the regular fingerprint query for distinct values.

File: src/metabase/models/field_values.clj [hunk 4]
```diff
@@ -106,36 +106,50 @@
...
   FieldValues. You most likely should not be using this directly in code outside of this namespace, unless it's for a
   very specific reason, such as certain cases where we fetch ad-hoc FieldValues for GTAP-filtered Fields.)"
   [field]
   (classloader/require 'metabase.db.metadata-queries)
   (try
-    (let [values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)]
-      (when (values-less-than-total-max-length? values)
-        values))
+    (let [distinct-values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)
+          values          (into [] (take-by-length total-max-length) distinct-values)]
```
Comment (noahmoss) [inline src/metabase/models/field_values.clj:141]: Maybe call this `limited-values` or `limited-distinct-values` or something? `values` feels like it should be a superset of `distinct-values` instead of a subset, but maybe that's just me

File: src/metabase/models/field_values.clj [hunk 5]
```diff
@@ -106,36 +106,50 @@
...
+      {:values          values
+       ;; has_more_values=true means the list of values we return is a subset of all possible values.
+       ;; has_more_values=true when:
+       ;; 1. the number of distinct values of this field exceeded [[metabase.db.metadata-queries/absolute-max-distinct-values-limit]]
```
Comment (noahmoss) [inline src/metabase/models/field_values.clj:145]: This line confused me at first when I read it, and tried to reconcile it with the `=` being used on L151.

My understanding is now: `metabase.db.metadata-queries/field-distinct-values` will return a maximum of `absolute-max-distinct-values-limit` fields. If the result is exactly equal to the limit, we err on the side of assuming there were more values we didn't read. Is that right? If so, could you clarify this somehow - maybe just add a comment explaining why `=` is being used in L151?

Just brainstorming now...we could change the query to have a limit of `absolute-max-distinct-values-limit + 1` but keep the original limit for the number of fields we actually store. Then we can use a `>` comparison and be very sure whether there are extra values in the underlying field. Maybe not worth the effort though.

**Review by noahmoss — APPROVED** [review]
Code looks reasonable, so stamping to unblock, with just some minor comments/questions. 
Although I don't 100% understand whether the concerns above about the overall approach have been resolved.

File: src/metabase/models/field_values.clj [hunk 6]
```diff
@@ -106,36 +106,50 @@
...
-                      0
-                      distinct-values)]
-    (u/prog1 (<= total-length total-max-length)
-      (log/debug (trs "Field values total length is > {0}." total-max-length)
-                 (if <>
-                   (trs "FieldValues are allowed for this Field.")
-                   (trs "FieldValues are NOT allowed for this Field."))))))
+
+(defn take-by-length
+  "Returns a transducer that takes a collection and returns a collection where the total length of all items in list is less than `max-length`."
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/field_values.clj:111]: It's not, it's just fun to write 😆.
I added a 2-arity with lazy-seq version and use it instead. 
but imma keep the transducer version in case we want to use it later.

File: src/metabase/models/field_values.clj (hunk 4 above)
Comment (qnkhuat) [reply src/metabase/models/field_values.clj:141]: I agree, `limited-distinct-values` makes much more sense

File: src/metabase/models/field_values.clj (hunk 5 above)
Comment (qnkhuat) [reply src/metabase/models/field_values.clj:145]: yea it's a bit tricky, I updated the docs to make it clearer. Could you have a second look?

File: src/metabase/models/field_values.clj [hunk 7]
```diff
@@ -106,36 +106,65 @@
...
   (try
-    (let [values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)]
-      (when (values-less-than-total-max-length? values)
-        values))
+    (let [distinct-values         ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)
+          limited-distinct-values (take-by-length total-max-length distinct-values)]
+      {:values          limited-distinct-values
+       ;; has_more_values=true means the list of values we return is a subset of all possible values.
+       :has_more_values (or
+                          ;; If the `distinct-values` has more elements than `limited-distinct-values`
```
Comment (qnkhuat) [inline src/metabase/models/field_values.clj:158]: @noahmoss the updated comment is here
